### PR TITLE
Add COS integration

### PR DESCRIPTION
**Pull request by weiiwang01**
Integrate COS into the charm by installing `chrony_exporter` from [Canonical's devops Launchpad](https://launchpad.net/~canonical-is-devops/+archive/ubuntu/chrony-charm) and adding support for the `grafana-agent` and implementing a new chrony Grafana dashboard.

It's important to note that, at present, the `cos_agent` charm library is incompatible with pydantic v2. Although a pull request (https://github.com/canonical/grafana-agent-operator/pull/156) has been submitted to fix this issue, the version of the `cos_agent` library currently used in this pull request is not yet available on charmhub. Additionally, the `grafana-agent` in charmhub does not support `ubuntu@24.04`, thus it cannot be deployed from charmhub.

Screenshot 2024-08-01 at 01 22 04

### CODE REVIEW

**Comment by cbartz** [comment]
@weiiwang01 Would you mind adding/using the pull request template and adding labels for the complexity of the PR? That would be great and make it easier to review the PR, thanks.

**Review by arturo-seijas — DISMISSED** [review]

**Review by cbartz — DISMISSED** [review]
LGTM, minor comments

**Comment by weiiwang01** [comment]
> @weiiwang01 Would you mind adding/using the pull request template and adding labels for the complexity of the PR? That would be great and make it easier to review the PR, thanks.

I have added the label to the pull request. Regarding the template, I think it might contain some outdated information, for example, [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619), and could add too much noise to the pull request description. Therefore, I have opted out of using the template temporarily.

File: lib/charms/grafana_agent/v0/cos_agent.py [hunk 1]
```diff
@@ -0,0 +1,809 @@
...
+import logging
+from collections import namedtuple
+from itertools import chain
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Callable, ClassVar, Dict, List, Optional, Set, Tuple, Union
+
+try:
+    import pydantic.v1 as pydantic
+except ImportError:
+    import pydantic
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply lib/charms/grafana_agent/v0/cos_agent.py:219]: Yes, I think we might need to wait until https://github.com/canonical/grafana-agent-operator/pull/156 is merged before merge this pull request.

File: requirements.txt [hunk 2]
```diff
@@ -3,3 +3,4 @@ jsonschema==4.23.0
 ops==2.15.0
 pydantic==2.8.2
 rpds-py==0.18.1
+cosl==0.0.13
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply requirements.txt]: Updated, thanks!

File: src/grafana_dashboards/chrony.json [hunk 3]
```diff
@@ -0,0 +1,1379 @@
...
+        "tooltip": {
+          "mode": "single",
+          "sort": "none"
+        }
+      },
+      "targets": [
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "P5887CE5F53B228A3"
```
Comment (yanksyoon) [inline src/grafana_dashboards/chrony.json:126]: I think some UIDs are intended to be: `"uid": "${prometheusds}"`

File: src/grafana_dashboards/chrony.json [hunk 4]
```diff
@@ -0,0 +1,1379 @@
...
+          "refId": "A"
+        }
+      ],
+      "title": "NTP Packets",
+      "type": "timeseries"
+    },
+    {
+      "datasource": {
+        "type": "prometheus",
+        "uid": "P5887CE5F53B228A3"
```
Comment (yanksyoon) [inline src/grafana_dashboards/chrony.json:140]: Here as well :D

File: src/grafana_dashboards/chrony.json [hunk 5]
```diff
@@ -0,0 +1,1379 @@
...
+    "from": "now-24h",
+    "to": "now"
+  },
+  "timepicker": {},
+  "timezone": "",
+  "title": "chrony",
+  "uid": "c252be85-471b-41c9-b1cb-f0e1b8206593",
+  "version": 4,
+  "weekStart": ""
+}
```
Comment (yanksyoon) [inline src/grafana_dashboards/chrony.json:1379]: Minor: missing newline :D

**Review by yanksyoon — DISMISSED** [review]
Minor nits, LGTM!

File: src/grafana_dashboards/chrony.json [hunk 6]
```diff
@@ -0,0 +1,1379 @@
...
+      },
+      "id": 22,
+      "panels": [],
+      "title": "Server Stats",
+      "type": "row"
+    },
+    {
+      "datasource": {
+        "type": "prometheus",
+        "uid": "P5887CE5F53B228A3"
```
Comment (amandahla) [inline src/grafana_dashboards/chrony.json:48]: Shouldnt be something like "uid": "${prometheusds}"?

File: src/grafana_dashboards/chrony.json [hunk 7]
```diff
@@ -0,0 +1,1379 @@
...
+    "list": []
+  },
+  "time": {
+    "from": "now-24h",
+    "to": "now"
+  },
+  "timepicker": {},
+  "timezone": "",
+  "title": "chrony",
+  "uid": "c252be85-471b-41c9-b1cb-f0e1b8206593",
```
Comment (amandahla) [inline src/grafana_dashboards/chrony.json:1376]: uid: null ?

File: src/grafana_dashboards/chrony.json [hunk 8]
```diff
@@ -0,0 +1,1379 @@
...
+  "tags": [],
+  "templating": {
+    "list": []
+  },
+  "time": {
+    "from": "now-24h",
+    "to": "now"
+  },
+  "timepicker": {},
+  "timezone": "",
```
Comment (amandahla) [inline src/grafana_dashboards/chrony.json:1374]: Usually is UTC.

File: src/grafana_dashboards/chrony.json [hunk 9]
```diff
@@ -0,0 +1,1379 @@
...
+  "title": "chrony",
+  "uid": "c252be85-471b-41c9-b1cb-f0e1b8206593",
+  "version": 4,
```
Comment (amandahla) [inline src/grafana_dashboards/chrony.json:1377]: minor: version 1

File: src/grafana_dashboards/chrony.json [hunk 10]
```diff
@@ -0,0 +1,1379 @@
...
+  "templating": {
+    "list": []
+  },
+  "time": {
+    "from": "now-24h",
+    "to": "now"
+  },
+  "timepicker": {},
+  "timezone": "",
+  "title": "chrony",
```
Comment (amandahla) [inline src/grafana_dashboards/chrony.json:1375]: What about Chrony Operator?

File: src/grafana_dashboards/chrony.json (hunk 3 above)
Comment (weiiwang01) [reply src/grafana_dashboards/chrony.json:126]: Indeed, updated, thanks!

File: src/grafana_dashboards/chrony.json (hunk 4 above)
Comment (weiiwang01) [reply src/grafana_dashboards/chrony.json:140]: Updated, thanks!

File: src/grafana_dashboards/chrony.json (hunk 6 above)
Comment (weiiwang01) [reply src/grafana_dashboards/chrony.json:48]: Yes, updated, thanks!

File: src/grafana_dashboards/chrony.json (hunk 7 above)
Comment (weiiwang01) [reply src/grafana_dashboards/chrony.json:1376]: Updated, thanks!

File: src/grafana_dashboards/chrony.json (hunk 8 above)
Comment (weiiwang01) [reply src/grafana_dashboards/chrony.json:1374]: Added, thanks!

File: src/grafana_dashboards/chrony.json (hunk 9 above)
Comment (weiiwang01) [reply src/grafana_dashboards/chrony.json:1377]: Updated, thanks!

File: src/grafana_dashboards/chrony.json (hunk 10 above)
Comment (weiiwang01) [reply src/grafana_dashboards/chrony.json:1375]: Yes, that's a good idea, update the title and the description, thanks!

File: src/grafana_dashboards/chrony.json (hunk 5 above)
Comment (weiiwang01) [reply src/grafana_dashboards/chrony.json:1379]: Updated, thanks!

**Review by amandahla — DISMISSED** [review]
LGTM

**Review by arturo-seijas — APPROVED** [review]

**Review by cbartz — APPROVED** [review]

**Pull request merged**